### PR TITLE
Fix failure on aliases and datastreams

### PR DESCRIPTION
**Pull request by LucasStinchcombe**
There is also some dead code in [OpensearchClient.java](https://github.com/aiven/opensearch-connector-for-apache-kafka/blob/v1.0.1/src/main/java/io/aiven/kafka/connect/opensearch/OpensearchClient.java#L126) which this change causes, but it seems like from [integration tests](https://github.com/aiven/opensearch-connector-for-apache-kafka/blob/6e2bef7dfc645c3d8728a2c15c25ea0acf303b7d/src/integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchClientIT.java#L62) it seems it should stay idempotent.

Basically indexes, aliases, and datastreams all return true on
```
client.indices().exists(new GetIndexRequest(index), RequestOptions.DEFAULT);
```

So checking beforehand and not trying to create an index should solve the exception raised in #85.
A caveat for datastreams: a datastream is not recognized as an index with an index template alone, it requires the datastream to be intialized via `PUT _data_stream/the-datastream`.

### CODE REVIEW

File: src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java [hunk 1]
```diff
@@ -157,8 +157,10 @@ protected String convertTopicToIndexName(final String topic) {
 
     private void ensureIndexExists(final String index) {
         if (!indexCache.contains(index)) {
-            LOGGER.info("Create index {}", index);
-            client.createIndex(index);
+            if (!client.indexExists(index)) {
```
Comment (reta) [inline src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java:160]: @LucasStinchcombe thanks for the fix, but it won't address the problem at large: there are still posibility for index to be created in between of `indexExists` and `createIndex` calls.  I think we should be focusing on the fact that index may exists during the `createIndex`: should that be fatal or non-fatal condition?

Comment (LucasStinchcombe) [reply src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java:160]: Sure, if we want it to be non-fatal at `createIndex`, then following suite with current implementation, we would just ignore two additional cases in [createIndex:L124](https://github.com/aiven/opensearch-connector-for-apache-kafka/blob/main/src/main/java/io/aiven/kafka/connect/opensearch/OpensearchClient.java#L124). Although I personally think that checking strings in return error is brittle. I'll check what the values of `RESOURCE_EXISTS_AS_ALIAS` and `RESOURCE_EXISTS_AS_DATASTREAM` should be.

```java
if (!(e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)
    | | e.getMessage().contains(RESOURCE_EXISTS_AS_ALIAS)
    | | e.getMessage().contains(RESOURCE_EXISTS_AS_DATASTREAM)))
{
  throw e;
}
return false;
```

Let me know what you think.

Comment (reta) [reply src/main/java/io/aiven/kafka/connect/opensearch/OpensearchSinkTask.java:160]: :+1:  Thanks @LucasStinchcombe , I think it would be concise with the way the error handling is implemented right now (even if it is not ideal), thanks again.

**Comment by reta** [comment]
LGTM, could you please add integration test cases to `integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchClientIT.java`, thank you

**Comment by LucasStinchcombe** [comment]
I've added a simple integration test for aliases which just checks that index creation returns `false` and doesn't throw exception when alias exists. I want to do the same for datastreams but couldn't find how to create an index template designated as a datastream.

Basically the equivalent of
```
PUT _index_template/sample-index-template                                       
{                                                                               
  "index_patterns": [                                                           
    "sample-datastream"                                                         
  ],                                                                            
  "data_stream": {},                                                            
  "priority": 100                                                               
}

PUT _data_stream/sample-datastream
```

I was looking at [PutIndexTemplateRequest](https://javadoc.io/static/org.opensearch.client/opensearch-rest-high-level-client/2.3.0/org/opensearch/client/indices/PutIndexTemplateRequest.html) for that but it has no methods to tell it to apply to datastreams. Only a method [includeDataStreams](https://github.com/opensearch-project/OpenSearch/blob/6ca3278f96ab60312e5e039922ac5206341a3393/server/src/main/java/org/opensearch/action/IndicesRequest.java#L63) inherited from Indices Request which always returns `false`.

File: src/integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchClientIT.java [hunk 2]
```diff
@@ -65,6 +68,29 @@ void createIndexDoesNotCreateAlreadyExistingIndex() {
...
+        } catch (final OpenSearchStatusException | IOException e) {
+            throw e;
+        }
+
+        assertFalse(opensearchClient.createIndex("alias_index"));
+    }
+
+    @Test
+    void createIndexDoesNotCreateAlreadyExistingDatastream() {
+        // TODO create datastream "datastream_index"
```
Comment (reta) [inline src/integration-test/java/io/aiven/kafka/connect/opensearch/OpensearchClientIT.java:90]:
```
    @Test
    void createIndexDoesNotCreateWhenAliasExists() throws Exception {
        final var config = new OpensearchSinkConnectorConfig(getDefaultProperties());
        final OpensearchClient tmpClient = new OpensearchClient(config);

        try {
            final ComposableIndexTemplate template = new ComposableIndexTemplate(
                    Arrays.asList("my-data-stream", "index-logs-*"),
                    null,
                    null,
                    100L,
                    null,
                    null,
                    new DataStreamTemplate());  
            PutComposableIndexTemplateRequest request = new PutComposableIndexTemplateRequest();
            request.name("data-stream-template");
            request.indexTemplate(template);
            
            tmpClient.client.indices().putIndexTemplate(request, RequestOptions.DEFAULT);
            tmpClient.client.indices().createDataStream(new CreateDataStreamRequest("my-data-stream"), RequestOptions.DEFAULT);
        } catch (final OpenSearchStatusException | IOException e) {
            throw e;
        }

        assertFalse(opensearchClient.createIndex("index-logs-0"));
    }
```

**Comment by reta** [comment]
> I was looking at [PutIndexTemplateRequest](https://javadoc.io/static/org.opensearch.client/opensearch-rest-high-level-client/2.3.0/org/opensearch/client/indices/PutIndexTemplateRequest.html) for that but it has no methods to tell it to apply to datastreams. Only a method [includeDataStreams](https://github.com/opensearch-project/OpenSearch/blob/6ca3278f96ab60312e5e039922ac5206341a3393/server/src/main/java/org/opensearch/action/IndicesRequest.java#L63) inherited from Indices Request which always returns `false`.

Thanks @LucasStinchcombe , posted the test case sample for you (this is different API, putIndexTemplate / PutComposableIndexTemplateRequest).

**Comment by LucasStinchcombe** [comment]
Thanks for the code sample. I noticed that I had a mistake in the other integration test, and yet CI pipeline didn't complain.
Are integration tests run before release?

**Comment by reta** [comment]
> Thanks for the code sample. I noticed that I had a mistake in the other integration test, and yet CI pipeline didn't complain. Are integration tests run before release?

Thanks @LucasStinchcombe , I suspect that integration tests are run against each pull request

**Review by reta — APPROVED** [review]
